File: src/player.py
from src.board import Board, Move
from simple_websocket import Client, ConnectionClosed
from time import time_ns
from enum import Enum

from src.consts import CONFIG

PlayerState = Enum("PlayerState", ["WAITING", "CALCULATING", "STOPPED"])


class Player():
    def __init__(self, ws: Client) -> None:
        self.get_full_position = lambda: ""
        self.time_last_move = time_ns()
        self.is_ready = False
        self.has_crashed = False
        self.ws = ws

        self.board = Board()

    def set_ready(self):
        self.is_ready = True
        self.state = PlayerState.WAITING

    def start_calculating(self):
        string = f"start position {self.get_full_position()}"
        self.ws.send(string)
        self.state = PlayerState.CALCULATING

    def send_illegal_move(self):
        # TODO send illegal penalties
        self.ws.send("illegal")
        self.state = PlayerState.STOPPED

    def send_rules(self):
        s = 'rules '
        for rule in CONFIG["rules"]:
            s += f"{rule} {CONFIG['rules'][rule]} "
        self.ws.send(s)

    def stop(self):
        self.state = PlayerState.STOPPED

    def play_move(self, move: Move):
        has_played_successfully = self.board.push(move)

        while time_ns() - self.time_last_move > 1000 / CONFIG["rules"]["pps"]:
            pass

        if has_played_successfully:
            self.time_last_move = time_ns()
            if self.board.has_lost:
                self.state = PlayerState.STOPPED
                return

            self.state = PlayerState.WAITING
        else:
            self.send_illegal_move(self.ws)

    def handle_message(self, message: str):
        words = message.split(" ")
        command = words[0]

        match command:
            case "info":
                self.send_rules()
            case "readyok":
                self.is_ready = True
                self.state = PlayerState.WAITING
            case "suggestion":
                pass
            case "bestmove":
                try:
                    move = Move.from_string(" ".join(words[1:]))
                except ValueError:
                    self.send_illegal_move()
                    return

                self.play_move(move)
                self.state = PlayerState.STOPPED

            case "quit":
                self.ws.send("quit ok")
                self.ws.close()
            case "crash":
                self.has_crashed = True

            case _:
                self.ws.send("error invalid command")

    def reset(self):
        self.board = Board()
        self.state = PlayerState.WAITING

File: src/player.ts
import { rules } from "../config.json";

import { Board, Move } from "./board";
import * as console from "./logging";

const debugMode = process.argv.includes("--debug");

const rulesString = (() => {
	let string = "";
	for (const [key, value] of Object.entries(rules)) {
		string += `${key.replace(
			/[A-Z]/g,
			letter => `-${letter.toLowerCase()}`
		)} ${value} `;
	}
	return string;
})();

export default class Player {
	private ws: WebSocket;
	private infos: {
		name: string;
		version: string;
		author: string;
	} = {
		name: "",
		version: "",
		author: ""
	};
	private board = new Board();

	public id: string;
	public onMove: () => void = () => {};
	public getFullPosition: () => string = () => "";

	public isReady = false;
	constructor(ws: WebSocket, uuid: string) {
		this.ws = ws;
		this.id = uuid;
	}

	get position() {
		return this.board.toString();
	}

	send(msg: string) {
		if (debugMode) {
			console.debug(
				`Message to ${this.id.split("-")[0]}: ${
					msg.length > 15 ? msg.slice(0, 15) + "..." : msg
				}`
			);
		}

		this.ws.send(msg);
	}

	crash(msg: string) {
		this.send(`error ${msg}`);
		this.ws.close();
	}

	set info(argv: string[]) {
		let parsedInfos = { name: "", version: "", author: "" };
		for (let i = 0; i < argv.length; i += 2) {
			let key = argv[i];
			let value = argv[i + 1];
			if (key === "name" || key === "version" || key === "author") {
				parsedInfos[key] = value;
			}
		}
		this.infos = parsedInfos;
	}

	get info(): { name: string; version: string; author: string } {
		return this.infos;
	}

	get hasLost() {
		return this.board.hasLost;
	}

	sendRules() {
		this.send(`rules ${rulesString}`);
	}

	resetBoard() {
		this.board = new Board();
	}

	startCalculation() {
		const fullPosition = this.getFullPosition();

		this.send(`start position ${fullPosition}`);
		setTimeout(() => {
			this.send("play");
		}, 1000 / rules.pps);
	}

	illegalMoveProcedure() {
		if (debugMode) {
			console.debug(`Illegal move by ${this.id.split("-")[0]}`);
		}

		// TODO: Illegal move logic
	}

	playMove(move: Move | null) {
		if (move === null) {
			this.illegalMoveProcedure();
			return;
		}

		const hasPlayedSuccessfully = this.board.push(move);

		if (!hasPlayedSuccessfully) {
			this.illegalMoveProcedure();
		}
	}

	command(msg: string) {
		let argv = msg.split(" ");
		let command = argv.shift();

		switch (command) {
			case "info":
				this.sendRules();
				this.info = argv;
				break;
			case "readyok":
				this.isReady = true;
				break;
			case "suggestion":
				break;
			case "bestmove":
				this.playMove(Move.fromString(argv.join(" ")));
				this.onMove();
				break;
			case "quit":
				this.send("quit ok");
				this.ws.close();
				break;
			case "crash":
				break;
			default:
				this.send("error invalid-command");
				break;
		}
	}
}
